### DISCUSSION TITLE
Sending emails  ✉️

### DISCUSSION THREAD

**Discussion by mbrookson** [blitz-js/blitz#798] [Ideas]
Thought of something that could be a cool and useful feature/recipe... Emails. ✉️

This is both an idea and topic for discussion. How are other people handling sending of emails in their Blitz apps? 🙂 💡 

I've got a couple of options that I'm considering and not sure on the best approach.
- Use an email provider like SendGrid or SendInBlue to handle everything, but comes with more limits and higher cost.
- Use something like AWS SES for lower level control of email, but have to handle template creation manually which is more of a pain in some ways but don't need all the extra stuff that comes with the providers above.

For option 1, generally you create email templates within their system and use their APIs and/or SDKs to send them etc.

For option 2 I've seen a couple of projects that could help with this.
- https://github.com/leemunroe/responsive-html-email-template - Pure static html email templates. Need something more than this with dynamic templating.
- https://github.com/chromakode/react-html-email - Not been updated for years but seems cool to help write emails.
- https://reactjs.org/docs/react-dom-server.html - Could purely just use `react-dom/server` to render email templates from React components.

Emails are definitely something that are often required in a full stack application. I'm going to keep playing around within my own app and see what I come up with.

**Comment by igeligel** [comment]
You can also check https://mjml.io

I am using this personally + AWS SES. Maybe it would make sense here to have a common API as kind of a plugin and write different API wrappers.

**Reply by mbrookson** [reply]
How are you sending to SES? Directly from your Blitz API function? As @flybayer mentions, a queue in between would be preferable so your application isn't held up waiting for SES.

**Reply by igeligel** [reply]
You can of course integrate a queue (Amazon SQS for example) in there. In my personal Next.js/express application, I never had a problem to send an email directly using the API. Amazon is scaling by itself basically.

**Comment by flybayer** [comment]
Good question @mbrookson! 

I highly recommend using https://postmarkapp.com instead of doing templating inside your app. I've found it to be a lot nicer for other business folks to be able to update email templates without requiring a developer.

Also, you want send emails as a background process so they don't block the user's network request. If deploying Blitz serverless, you need to set up a queue and send email events to the queue. Or instead of all this, just send an API request to Postmark/Sendgrid :)

That said, I could totally see having a blitz email plugin for folks wanting to render their own template.

**Reply by mbrookson** [reply]
It's just a personal project so only going to be me editing the templates. However I still think it'd be much easier to be able to edit templates in a GUI rather than fiddling with CSS for email clients. Also I want as little infrastructure to manage as possible.

Forgive my ignorance, but you mention setting up a queue and I'm not fully clued up on how to implement this. I imagine it's pretty simple to set up SNS to directly trigger SES as it's all within the AWS ecosystem. Do you know what other ways you could achieve it though? What if you're not using AWS? What should be listening to the queue to actually send an email? Do you need a separate non-serverless server running to handle queue messages?

https://docs.bullmq.io/ looks pretty cool. I think the process with this is:
- Create your queue infrastructure. Think this library requires Redis.
- Put a message onto the queue
- Create a 'Worker' which receives queue messages and triggers a callback, where you could call out to SES.

However with the above flow, I'm not sure how/if the workers could work with serverless? Again, it's a question of what/where is that listener.

**Reply by mbrookson** [reply]
Also thinking about it, even if just calling a third-party API like SendGrid, isn't it still better to trigger that call from a queue so your own backend doesn't depend on a third-party API?

**Reply by flybayer** [reply]
For serverless, you need to use native AWS events inside their ecosystem (you can't use Vercel and do this), or use a queue that can trigger any HTTP endpoint. I've used https://www.cloudamqp.com for this with some success.

If you are sending emails with an API call, then you do not need a queue for that.

**Reply by mbrookson** [reply]
Yeah makes sense.

You wouldn't use a queue even if you're making the API from your own API?

**Reply by flybayer** [reply]
Nope! The API call will take, what, maybe 100ms? Making an API call to put the item in your queue would take around the same time.

Whereas sending an email via STMP would take significantly longer.

**Reply by mbrookson** [reply]
Yeah I get what you're saying. However, I was also thinking more around failures and retries. If a third-party's API is down or there is some sort of connectivity issue, or even misconfiguration on your own end maybe, you could avoid failed emails by using a queue/job workflow that retries and moves the responsibility away from your own API. You could build retry logic inline but obviously that could then slow down the requests a lot.

You'd hope a service like SendGrid would be pretty reliable so maybe it's not something to worry about. I'm just always cautious about relying on third-party APIs to work all the time 🤷‍♂️ I suppose that's also a lot of added complexity for potentially not a lot of value though.

I guess with critical emails, like an email verification which is required, you could also prompt the user and give them a link to retry if they didn't receive the expect email.

**Reply by flybayer** [reply]
**Yeah that's a very good point. Using a queue even for API calls would definitely be the most robust.** 

I guess depends on your app and business requirements. I have production apps running that just make API calls without a queue and it has been fine. But I can think of a couple situations where something would go wrong and something would be missed. Usually in this case, someone would contact customer support and get it resolved.

**Reply by Jarrodsz** [reply]
Serious need mjml templating inside blitz. Rails had actionMailer and easily provides extending it. I have all my email templates in rails and could easy port them to blitz.

Any suggestions on how to implement this?

How about an actionMailer like implementation for blitz that easily allow different templates and mail providers with separate documentation on blitz docs

**Reply by flybayer** [reply]
@Jarrodsz yeah we totally want to add support for this. We just only have so much bandwidth. I would be more than happy to work with someone who can dedicate time to work on it and figure out the best solution.

**Reply by roqkabel** [reply]
> @Jarrodsz yeah we totally want to add support for this. We just only have so much bandwidth. I would be more than happy to work with someone who can dedicate time to work on it and figure out the best solution.

@flybayer  I would be glad to work on this, I have been a fan of rails Action Mailer, and I think doing same for blitz could be a great feature for devs.
